src/panels/lovelace/mixins/element-click-mixin.js

```diff
@@ -10,13 +10,71 @@ import computeStateName from '../../../common/entity/compute_state_name';
  * @appliesMixin NavigateMixin
  */
 export default dedupingMixin(superClass => class extends NavigateMixin(EventsMixin(superClass)) {
-  handleClick(hass, config) {
-    const tapAction = config.tap_action || 'more-info';
+
+  registerMouse(config, enableHold = false) {
+    this.mouseHeld = false;
+    this.mouseHoldTimer = null;
+
+    let reset = () => {
+      clearTimeout(this.mouseHoldTimer);
+      this.mouseHeld = false;
+    };
+
+    let click = () =>  {
+      let held = this.mouseHeld;
+      this.mouseHeld = false;
+      if(held) return;
+      this.handleClick(this.hass, config, false);
+    };
+
+    let down = () => {
+      this.mouseHoldTimer = setTimeout(() => {
+        this.mouseHeld = true;
+        this.handleClick(this.hass, config, true);
+      }, 300);
+    };
+
+    let up = () => {
+      clearTimeout(this.mouseHoldTimer);
+      let moreInfoDialog = document
+        .querySelector('home-assistant')
+        .$$('ha-more-info-dialog');
+      if (moreInfoDialog) {
+        setTimeout(() => moreInfoDialog.removeAttribute('modal'), 0);
+      }
+    };
+
+    this.addEventListener('click', click);
+
+    this.addEventListener('mouseout', reset);
+    this.addEventListener('touchmove', reset);
+    document.addEventListener('mouseup', up);
+    document.addEventListener('touchend', up);
+
+    if (config.hold_action) {
+      this.addEventListener('mousedown', down);
+      this.addEventListener('touchstart', down);
+    }
+  }
+
+  handleClick(hass, config, held=false) {
+    let tapAction;
+    if(held) {
+      tapAction = config.hold_action || 'none';
+    } else {
+      tapAction = config.tap_action || 'more-info';
+    }
     if (tapAction === 'none') return;
 
     switch (tapAction) {
       case 'more-info':
         this.fire('hass-more-info', { entityId: config.entity });
+        if(held) {
+          let moreInfoDialog = document
+            .querySelector('home-assistant')
+            .$$('ha-more-info-dialog');
+          moreInfoDialog.setAttribute('modal', true);
+        }
```

Comment:
Once the mouse button is released after a `hold`, a `mouseup` or `click` event is fired. If the more-info-dialog was opened as `hold_action`, that will immediately close it again. Therefore the dialog is given the attribute `modal`, which means it's not closed by click outside the dialog. The `modal` attribute is removed on `mouseup` above (line 43).